### PR TITLE
Avoid losing errors from launcher

### PR DESCRIPTION
**Pull request by propensive**

### CODE REVIEW

File: src/core/data.scala [hunk 1]
```diff
@@ -336,14 +336,14 @@ object BspConnectionManager {
...
-                    broken: Promise[Unit],
+                    broken: Promise[Exception],
                     launcher: Future[Unit])
             extends AutoCloseable {
 
     lazy val errReader = new BufferedReader(new InputStreamReader(err))
 
     override def close(): Unit = {
-      broken.success(())
+      broken.success(new Exception("Closed"))
```
Comment (odisseus) [inline src/core/data.scala:346]: When a connection is being closed, that doesn't necessarily mean something has failed. If I'm not mistaken, Bloop will close connections that have been unused for some time.

File: src/core/data.scala [hunk 2]
```diff
@@ -418,10 +426,10 @@ object BspConnectionManager {
       )
     }).map {
       case SuccessfulRun => ()
-      case failure       => throw new Exception(s"Launcher failed: $failure")
+      case failure       => broken.success(new Exception(s"Launcher failed: $failure")); throw new Exception()
```
Comment (odisseus) [inline src/core/data.scala:429]: Consider using [`Promise.failure(Throwable)`](https://www.scala-lang.org/api/current/scala/concurrent/Promise.html#failure(cause:Throwable):Promise.this.type). Signaling a failure with a successful result of type `Exception` is unusual, to say the least.

**Review by odisseus — CHANGES_REQUESTED** [review]

File: src/core/data.scala (hunk 1 above)
Comment (propensive) [reply src/core/data.scala:346]: Yes, though I needed to put *something* into the promise, now that I've given it the capability of carrying an exception. For now, I don't mind seeing "Closed" if it means I also get to see the errors from the launcher, which I'm not currently seeing.

File: src/core/data.scala (hunk 2 above)
Comment (propensive) [reply src/core/data.scala:429]: Yes, you're right. It was a straight swap from a unit to an exception. I'm not convinced that using promises for this is a good idea in general (though I may be later), so I see a lot of this code changing again before too long.

**Pull request closed without merging**